x/mongo/driver/operation/hello.go

```diff
@@ -113,6 +121,430 @@ func (h *Hello) Result(addr address.Address) description.Server {
+	envNameVercel    = "vercel"
+)
+
+// getFaasEnvName parses the FaaS environment variable name and returns the
+// corresponding name used by the client. If none of the variables or variables
+// for multiple names are populated the client.env value MUST be entirely
+// omitted.
+func getFaasEnvName() string {
+	envVars := []string{
+		envVarAWSExecutionEnv,
+		envVarAWSLambdaRuntimeAPI,
+		envVarFunctionsWorkerRuntime,
+		envVarKService,
+		envVarFunctionName,
+		envVarVercel,
+	}
+
+	// If none of the variables are populated the client.env value MUST be
+	// entirely omitted.
+	names := make(map[string]struct{})
+
+	for _, envVar := range envVars {
+		if os.Getenv(envVar) == "" {
+			continue
+		}
+
+		var name string
+
+		switch envVar {
+		case envVarAWSExecutionEnv, envVarAWSLambdaRuntimeAPI:
+			name = envNameAWSLambda
+		case envVarFunctionsWorkerRuntime:
+			name = envNameAzureFunc
+		case envVarKService, envVarFunctionName:
+			name = envNameGCPFunc
+		case envVarVercel:
+			name = envNameVercel
+		}
+
+		names[name] = struct{}{}
+		if len(names) > 1 {
+			// If multiple names are populated the client.env value
+			// MUST be entirely omitted.
+			names = nil
+
+			break
+		}
+	}
+
+	for name := range names {
+		return name
+	}
+
+	return ""
+}
+
+// appendStringElement appends the key and value to dst. maxLen is the maximum
+// number of bytes that can be appended to dst.
+func appendStringElement(dst []byte, key, value string, maxLen int32) (int32, []byte) {
+	originalLen := int32(len(dst))
+	if int32(len(key)+len(value))+stringElementSize > maxLen {
+		return 0, dst
+	}
+
+	dst = bsoncore.AppendStringElement(dst, key, value)
+
+	bytesWritten := int32(len(dst)) - originalLen
+	return bytesWritten, dst
+}
+
+// appendInt32Element appends the key and value to dst. maxLen is the maximum
+// number of bytes that can be appended to dst.
+func appendInt32Element(dst []byte, key string, value int32, maxLen int32) (int32, []byte) {
+	originalLen := int32(len(dst))
+	if int32(len(key))+int32ElementSize > maxLen {
+		return 0, dst
+	}
+
+	dst = bsoncore.AppendInt32Element(dst, key, value)
+
+	bytesWritten := int32(len(dst)) - originalLen
+	return bytesWritten, dst
+}
+
+// appendClientAppName appends the application name to dst. It builds the
+// application sub-document key-by-key, checking the length of the document
+// after each key. If the document exceeds the maximum length, the key is
+// omitted and the document is ended and returned. MaxLen is the maximum number
+// of bytes that can be appended to dst.
+func appendClientAppName(dst []byte, name string, maxLen int32) (int32, []byte, error) {
+	const key = "application"
+
+	// Validate input parameters.
+	if dst == nil || maxLen <= 0 {
+		return 0, dst, nil
+	}
+
+	// If the size of the embedded document with the "application" key is
+	// greater than the maximum length, then the document is too long and
+	// we should terminate.
+	if embeddedDocumentSize+int32(len(key)) > maxLen {
```

Comment:
Can we fall back by truncating and restoring the document length with `bsoncore.ReadLength`/`bsoncore.UpdateLength`?